pkg/output/kafka/exporter.go

```diff
@@ -0,0 +1,108 @@
+package kafka
+
+import (
+	"context"
+	"errors"
+	"github.com/IBM/sarama"
+	"github.com/ethpandaops/xatu/pkg/observability"
+	"github.com/ethpandaops/xatu/pkg/proto/xatu"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+type ItemExporter struct {
+	config *Config
+	log    logrus.FieldLogger
+	client sarama.SyncProducer
+}
+
+func NewItemExporter(name string, config *Config, log logrus.FieldLogger) (ItemExporter, error) {
+
+	producer, err := NewSyncProducer(config)
+
+	if err != nil {
+		log.
+			WithError(err).
+			WithField("output_name", name).
+			WithField("output_type", SinkType).
+			Error("Error while creating the Kafka Client")
+		return ItemExporter{}, err
+	}
+
+	return ItemExporter{
+		config: config,
+		log:    log.WithField("output_name", name).WithField("output_type", SinkType),
+		client: producer,
+	}, nil
+}
+func (e ItemExporter) ExportItems(ctx context.Context, items []*xatu.DecoratedEvent) error {
+	_, span := observability.Tracer().Start(ctx, "KafkaItemExporter.ExportItems", trace.WithAttributes(attribute.Int64("num_events", int64(len(items)))))
+	defer span.End()
+
+	e.log.WithField("events", len(items)).Debug("Sending batch of events to Kafka sink")
+
+	if err := e.sendUpstream(ctx, items); err != nil {
+		e.log.
+			WithError(err).
+			WithField("num_events", len(items)).
+			Error("Failed to send events upstream")
+
+		span.SetStatus(codes.Error, err.Error())
+
+		return err
+	}
+
+	return nil
+}
+
+func (e ItemExporter) Shutdown(ctx context.Context) error {
+	return nil
+}
+
+const MAX_MSG_BYTE_SIZE = 1024 * 1024
+
+func (e *ItemExporter) sendUpstream(ctx context.Context, items []*xatu.DecoratedEvent) error {
+	msgs := make([]*sarama.ProducerMessage, 0, len(items))
+	msgByteSize := 0
+	for _, p := range items {
+		r, err := protojson.Marshal(p)
+		if err != nil {
+			return err
+		}
+		routingKey := sarama.StringEncoder(p.Event.Id)
+		eventPayload := sarama.StringEncoder(r)
+		m := &sarama.ProducerMessage{
+			Topic: e.config.Topic,
+			Key:   routingKey,
+			Value: eventPayload,
+		}
+		msgByteSize = m.ByteSize(2)
```

Comment:
Should this be `msgByteSize = msgByteSize + m.ByteSize(2)`?

Reply:
It would be just the `msgByteSize` for the following reasons

1. The sync producer is indeed an async producer beneath. So It wouldn't require us to check for the whole batch size (in bytes).
2. The only possible case, for the producer to enter into an infinite retry would be when `msgByteSize` > `max_message_bytes`. `msgByteSize > MAX_MSG_BYTE_SIZE` check would eliminate such a possibility. Further, a Dead Letter Queue is a nice addition to have in the future.